API: Polar: allow flipped y/rlims....

jklymak opened the pull request:
## PR Summary
Closes #11202
Closes #10704

## Update

Allow non-increasing ylimits:

  - [x] needs doc updates

```python

import numpy as np
import matplotlib.pyplot as plt
fig = plt.figure()

ax = fig.add_axes([0.1, 0.1, 0.8, 0.8], polar=True)
ax.plot(np.arange(30), np.arange(30) / 30. + 1)
ax.set_ylim(2, 1)  # out of order, and hence didn't draw properly.

plt.show()
```

![polar](https://user-images.githubusercontent.com/1562854/46260793-62216500-c49f-11e8-8a1c-28ac524922f7.png)

```
ax.set_rorigin(3.)
```

![polar2](https://user-images.githubusercontent.com/1562854/46260812-c2b0a200-c49f-11e8-8440-a92b0f49e885.png)


## OLD:

This adds a check for set_ylim in polar to make sure they are in order and swap and emit an error if they are not.  Because I had to overload `axes.set_ylim` I changed the argument handling a bit (hopefully correctly).  

Fixes a typo in `axes/_base.py`

```python

import numpy as np
import matplotlib.pyplot as plt

fig = plt.figure(dpi=120)

ax = fig.add_axes([0.1, 0.1, 0.8, 0.8], polar=True)

ax.set_theta_zero_location('N')
ax.set_theta_direction(-1)  # anti-clockwise

ax.set_ylim(90, -45)  # out of order, and hence didn't draw properly.  
ax.set_yticks(np.arange(-45, 90+0.1, 15))

plt.show()
```

## PR Checklist

- [x] Has Pytest style unit tests
- [x] Code is [Flake 8](http://flake8.pycqa.org/en/latest/) compliant
- [ ] New features are documented, with examples if plot related
- [ ] Documentation is sphinx and numpydoc compliant
- [ ] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [ ] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

<!--
Thank you so much for your PR!  To help us review your contribution, please
consider the following points:

- A development guide is available at https://matplotlib.org/devdocs/devel/index.html.

- Help with git and github is available at
  https://matplotlib.org/devel/gitwash/development_workflow.html.

- Do not create the PR out of master, but out of a separate branch.

- The PR title should summarize the changes, for example "Raise ValueError on
  non-numeric input to set_xlim".  Avoid non-descriptive titles such as
  "Addresses issue #8576".

- The summary should provide at least 1-2 sentences describing the pull request
  in detail (Why is this change required?  What problem does it solve?) and
  link to any relevant issues.

- If you are contributing fixes to docstrings, please pay attention to
  http://matplotlib.org/devel/documenting_mpl.html#formatting.  In particular,
  note the difference between using single backquotes, double backquotes, and
  asterisks in the markup.

We understand that PRs can sometimes be overwhelming, especially as the
reviews start coming in.  Please let us know if the reviews are unclear or
the recommended next step seems overly demanding, if you would like help in
addressing a reviewer's comments, or if you have been waiting too long to hear
back on your PR.
-->

WeatherGod commented:
Devil's advocate here: (just about) everywhere else we are able to flip the order of the limits (particular favorite is pressure-as-height). Why shouldn't we be allowed to do something similar for polar plots?

jklymak commented:
@WeatherGod Fine with me!  But someone has to rewrite it so it works, which it does not right now...

jklymak commented:
That makes sense.  I’m fine w an error and if someone wants to implement inverted axes they can do so in future

ImportanceOfBeingErnest commented:
I agree with @timhoffm here. You don't need to implement axis inversion here, but if you allow arbitrary order now, it'll be hard to allow for axis inversion via `set_*lim` in the future. 

The error might point towards a possible workaround for now. For the r axis: "To mimic an inverted r-axis plot data.max()-data and set the ticklabels accordingly.". For theta axis: "Use `set_theta_direction` instead."

jklymak commented:
Changed to a ValueError and changed the doc string to reflect thats what will happen.  

I don't know that we need to point out all the possible work arounds in an error message - they tend to be hard to describe in a line or two.  But if you had something pithy, that'd be fine.

timhoffm commented on lib/matplotlib/projections/polar.py:
```diff
@@ -1182,12 +1183,87 @@ def set_rorigin(self, rorigin):
-    def set_rlim(self, *args, **kwargs):
+    def set_rlim(self, bottom=None, top=None, emit=True, auto=False, **kwargs):
+        """
+        See `~.polar.PolarAxes.set_ylim`
```
See `~.polar.PolarAxes.set_ylim`. Additionally, you can pass *rmin*/*rmax* in place of *ymin*/*ymax*.

jklymak replied:
?? Not following this.  We are trying to deprecate rmin, ymin etc.  (Well, I don't agree with calling them bottom and top, but...)

timhoffm replied:
Thought to document the current state, even if it's deprecated (well, which should be mentioned then). But since this had no documentation at all so far. It's also ok to keep this a secret.

timhoffm commented on lib/matplotlib/projections/polar.py:
```diff
@@ -1182,12 +1183,87 @@ def set_rorigin(self, rorigin):
+            top = bottom[1]
+            bottom = bottom[0]
+        if bottom >= top:
+            raise ValueError('polar axes y/r-limits must be increasing; '
```
Do we have a convention if error messages should start with a capital letter?

timhoffm commented on lib/matplotlib/projections/polar.py:
```diff
@@ -1182,12 +1183,87 @@ def set_rorigin(self, rorigin):
+    def set_ylim(self, bottom=None, top=None, emit=True, auto=False,
+                 *, ymin=None, ymax=None):
+        """
+        Set the data limits for the radial-axis
```
Period at the end.

jklymak commented on lib/matplotlib/projections/polar.py:
```diff
@@ -1182,12 +1183,87 @@ def set_rorigin(self, rorigin):
+            The bottom limit (default: None, which leaves the bottom
+            limit unchanged).
+            The bottom and top ylims may be passed as the tuple
+            (`bottom`, `top`) as the first positional argument (or as
```
OK< changed in `_base.set_x/ylim` as well, where I copied the docstring from.

timhoffm commented on lib/matplotlib/projections/polar.py:
```diff
@@ -1182,12 +1183,87 @@ def set_rorigin(self, rorigin):
+            These arguments are deprecated and will be removed in a future
+            version.  They are equivalent to bottom and top respectively,
+            and it is an error to pass both *ymin* and *bottom* or
+            `ymax` and `top`.
```
`*ymax* and *top*`

timhoffm commented on lib/matplotlib/axes/_base.py:
```diff
@@ -3074,25 +3074,25 @@ def set_xlim(self, left=None, right=None, emit=True, auto=False,
 
         right : scalar, optional
-            The right xlim (default: None, which leaves the right limit
+            The right xlim (default: ``None``, which leaves the right limit
```
We don't have a consistent policy for None, True, False so far.

There are all styles present in our docs:
- plain
- literal: double backticks
- italics: starred

Personally, I find double backslashes too cumbersome to type and visually too broad in plain text for there short words. I just use `*None*`.

The [numpydoc](https://numpydoc.readthedocs.io/en/latest/format.html#docstring-standard) and [sphinx](http://www.sphinx-doc.org/en/master/usage/extensions/napoleon.html) examples just use plain text.

Just mentioning. Since we currently have no standards, I won't hold this up.

jklymak replied:
Happy to adopt whatever standard folks like the best, or to start a standard!  It was plain before, happy to put it back.

jklymak commented:
Added a small test for ValueError....

jklymak commented:
squashed and rebased

tacaswell commented:
It looks like this used to work with inverted axes, but we broke it (likely inadvertently) in the polar refactor.  I think we should look into fixing it before we make it raise.

jklymak commented:
I don’t see any reason inverting the axis is difficult mathematically.  I just assumed it wasn’t working on purpose.

tacaswell commented:
I just tested all they way back to 1.5 and this never actually worked (but silently kept going).

Sorry for the noise :sheep: .

jklymak commented:
OK, this seems to work fine...

```python

import numpy as np
import matplotlib.pyplot as plt
fig = plt.figure()

ax = fig.add_axes([0.1, 0.1, 0.8, 0.8], polar=True)
ax.plot(np.arange(30), np.arange(30) / 30. + 1)
ax.set_ylim(2, 1)  # out of order, and hence didn't draw properly.

plt.show()
```

![polar](https://user-images.githubusercontent.com/1562854/46260793-62216500-c49f-11e8-8a1c-28ac524922f7.png)

jklymak commented:
Rebased....

jklymak commented on lib/matplotlib/projections/polar.py:
```diff
@@ -444,7 +446,10 @@ def refresh(self):
         vmin, vmax = self.base.view_limits(vmin, vmax)
-        return mtransforms.nonsingular(min(0, vmin), vmax)
+        if vmax > vmin:
+            # this allows inverted r/y-lims
```
fixed.

jklymak replied:
Note that the old way of doing this for sorted `vmax > vmin`, did this `min(0, vmin)`.  I'm not sure I understand why, but didn't want to change back compatibility...

QuLogic replied:
It's there precisely for backwards compatibility. Before negative radii were accepted, the minimum was always 0, so this is here to ensure the limits remains 0 if you've only plotted positive radii, but is the actual minimum if you have negative radii.

jklymak commented on lib/matplotlib/projections/polar.py:
```diff
@@ -2,6 +2,7 @@
 import types
 
 import numpy as np
+import warnings
```
fixed...

QuLogic commented:
Something funny happens if you flip the radial limits after a draw:
```python
#!/usr/bin/env python

import numpy as np
import matplotlib.pyplot as plt

fig, (ax0, ax1, ax2) = plt.subplots(1, 3, subplot_kw=dict(polar=True))

ax0.plot([0, 1], [-1, 1])
ax0.scatter([0, 1], [-1, 1])

ax0.set_rmin(-1.5)
ax0.set_rmax(1.5)
ax0.set_title('Forward')

ax1.plot([0, 1], [-1, 1])
ax1.scatter([0, 1], [-1, 1])

ax1.set_rmin(1.5)
ax1.set_rmax(-1.5)
ax1.set_title('Backward')

ax2.plot([0, 1], [-1, 1])
ax2.scatter([0, 1], [-1, 1])

ax2.set_rmin(-1.5)
ax2.set_rmax(1.5)
ax2.set_title('Backward with draw')
fig.canvas.draw()
ax2.set_rmin(1.5)
ax2.set_rmax(-1.5)

plt.show()
```
![figure_1](https://user-images.githubusercontent.com/302469/48756087-b35f0100-ec65-11e8-82c6-8e30dd0fb336.png)

jklymak commented:
@QuLogic Agreed, but that predates this PR, and is a different issue. On master:

```python 
import numpy as np
import matplotlib.pyplot as plt

fig, (ax0,  ax2) = plt.subplots(1, 2, subplot_kw=dict(polar=True))

ax0.plot([0, 1], [-1, 1])
ax0.scatter([0, 1], [-1, 1])

ax0.set_rmin(-1.5)
ax0.set_rmax(1.5)
ax0.set_title('Forward')


ax2.plot([0, 1], [-1, 1])
ax2.scatter([0, 1], [-1, 1])

ax2.set_rmin(-1.5)
ax2.set_rmax(1.5)
ax2.set_title('draw; change rlims')
fig.canvas.draw()
ax2.set_rmin(0)
ax2.set_rmax(1.5)

plt.show()
```

![fliplim](https://user-images.githubusercontent.com/1562854/48786204-30f13400-ec9b-11e8-8c77-0d9b094262ae.png)

jklymak commented:
Typos fixed - this has two approvals and could be merged if anyone has a minute...
